### ISSUE TITLE
"sub_iter.strides(0)[0] == 0 INTERNAL ASSERT FAILED at Reduce.cuh" when doing .sum() on large cuda tensor

### ISSUE THREAD

**Issue by dizcza** [pytorch/pytorch#37583]
## 🐛 Bug

Calling `.sum()` on large cuda tensors throws the error.

## To Reproduce

```
import torch
t = torch.zeros(5, 14400, 14400).cuda()
!nvidia-smi
t.sum(dim=0)
```
shows the following nvidia info
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64.00    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   40C    P0    32W / 250W |   4663MiB / 16280MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```
and the traceback is
```
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-bbe24183337c> in <module>()
      2 t = torch.zeros(5, 14400, 14400).cuda()
      3 get_ipython().system('nvidia-smi')
----> 4 t.sum(dim=0)

RuntimeError: sub_iter.strides(0)[0] == 0 INTERNAL ASSERT FAILED at /pytorch/aten/src/ATen/native/cuda/Reduce.cuh:706, please report a bug to PyTorch. 
```

This behavior is reproduced with any `(n, 14400, 14400)` tensor for `n>=5`.

**Note**. I can easily create a (6, 14400, 14400) tensor - memory usage, in this case, is `5455MiB / 16280MiB`.

## Expected behavior
Returns a (14400, 14400) tensor

## Environment

PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.3
[pip3] pytorch-mighty==0.1.0
[pip3] torch==1.5.0+cu101
[pip3] torchfile==0.1.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect

## Additional context

Google colab, GPU instance.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel @heitorschueroff

**Comment by pbelevich** [comment]
Reproducible

**Comment by ngimel** [comment]
cc @zasdfgbnm - could any of your last changes cause it? Probably not, but asking just in case.

**Comment by zasdfgbnm** [comment]
@ngimel Does 1.5.0 include my last change?

**Comment by zasdfgbnm** [comment]
I don't think so, the release tracker issue opens at March 18, and my PR is merged at April 1

**Comment by gchanan** [comment]
putting 1.5.1 milestone until we verify that it's not a problem in 1.5.0

**Comment by ngimel** [comment]
It is a problem in 1.5.0 (according to original issue, verified in colab), and likely has been a problem for a long time, so not a regression.

**Comment by gchanan** [comment]
ok, in that case I'll leave the milestone label until we make a call one way or another (i.e. if the fix is simple).

**Comment by timothybrooks** [comment]
This error message occurs for me with pytorch 1.7.0 when calling sum on a large tensor on GPU. The tensor is of shape `torch.Size([8, 1, 128, 1024, 1024])` and the call was x.sum(dim=1).

```
RuntimeError: sub_iter.strides(0)[0] == 0 INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1603729062494/work/aten/src/ATen/native/cuda/Reduce.cuh":928, please report a bug to PyTorch. 
```

**Comment by zasdfgbnm** [comment]
reproduced https://github.com/pytorch/pytorch/issues/37583#issuecomment-720172838 's report
```python
>>> a = torch.zeros(8, 1, 128, 1024, 1024)
>>> a.cuda().sum(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: sub_iter.strides(0)[0] == 0 INTERNAL ASSERT FAILED at "/home/gaoxiang/pytorch-cublas/aten/src/ATen/native/cuda/Reduce.cuh":928, please report a bug to PyTorch. 
Exception raised from gpu_reduce_kernel at /home/gaoxiang/pytorch-cublas/aten/src/ATen/native/cuda/Reduce.cuh:928 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x5c (0x7fa4a1bdffac in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x1c49d60 (0x7fa4a3849d60 in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so)
frame #2: <unknown function> + 0x1c3a1ef (0x7fa4a383a1ef in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so)
frame #3: <unknown function> + 0x137bd28 (0x7fa4cfa9fd28 in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #4: at::native::sum_out(at::Tensor&, at::Tensor const&, c10::ArrayRef<long>, bool, c10::optional<c10::ScalarType>) + 0xeb (0x7fa4cfa93ebb in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #5: at::native::sum(at::Tensor const&, c10::ArrayRef<long>, bool, c10::optional<c10::ScalarType>) + 0x58 (0x7fa4cfa95138 in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #6: <unknown function> + 0x272cb39 (0x7fa4a432cb39 in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so)
frame #7: <unknown function> + 0x277f74a (0x7fa4a437f74a in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so)
frame #8: <unknown function> + 0x1751808 (0x7fa4cfe75808 in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #9: at::sum(at::Tensor const&, c10::ArrayRef<long>, bool, c10::optional<c10::ScalarType>) + 0xdc (0x7fa4cfe75cec in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #10: <unknown function> + 0x2b2cf84 (0x7fa4d1250f84 in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x2b2d31a (0x7fa4d125131a in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #12: <unknown function> + 0x199c6c8 (0x7fa4d00c06c8 in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #13: at::Tensor::sum(c10::ArrayRef<long>, bool, c10::optional<c10::ScalarType>) const + 0xdc (0x7fa4d00c0ecc in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x4c169f (0x7fa4d5a6769f in /home/gaoxiang/.local/lib/python3.8/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #26: __libc_start_main + 0xf2 (0x7fa507856152 in /usr/lib/libc.so.6)
```

**Comment by hszhao** [comment]
> This error message occurs for me with pytorch 1.7.0 when calling sum on a large tensor on GPU. The tensor is of shape `torch.Size([8, 1, 128, 1024, 1024])` and the call was x.sum(dim=1).
> 
> ```
> RuntimeError: sub_iter.strides(0)[0] == 0 INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1603729062494/work/aten/src/ATen/native/cuda/Reduce.cuh":928, please report a bug to PyTorch. 
> ```

PyTorch 1.4 works fine for me, while 1.6 and 1.7 failed.

**Comment by rossimattia** [comment]
Hello,

I have the same error when calling `x.sum(dim=0)` with `x` a `[1, 80, 2314, 3096]` GPU tensor. The tensor `x` is the result of some operations on another tensor. I experimented the same error with PyTorch 1.4, 1.5, 1.6, 1.7.

If I generate a tensor `y = torch.ones(x.size(), dtype=x.dtype)` and call `y.sum(dim=0)`, then I get no error. However, calling `x.contiguous().sum(dim=0)` triggers the error as well.

**Comment by ndronen** [comment]
any news about this? i'm seeing it with 1.7.1.

**Comment by JarneVdH** [comment]
Also 1.7.1 here.
It has probably to do with running out of memory.
When lowering the batch size for example (or any other change which reduces the tensor within sum()), the issue is gone.

**Comment by EdwardJB** [comment]
> Also 1.7.1 here.
> It has probably to do with running out of memory.
> When lowering the batch size for example (or any other change which reduces the tensor within sum()), the issue is gone.

I'm seeing the same on 1.8.0 under WSL. I think you're right about memory, but the strange thing is it only happens if I use more than _exactly_ 50% of VRAM.

**Comment by eduardo4jesus** [comment]
This error just happened to me on `1.9.0` :/

**Comment by eduardo4jesus** [comment]
To have an idea on how often this happens to me, here is the log as I was measuring the runtime on my research project.

```
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (64, 1, 96, 224)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (64, 1, 128, 192)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (64, 1, 128, 208)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (64, 1, 128, 224)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (64, 1, 256, 144)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (64, 1, 256, 160)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (64, 1, 256, 176)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (64, 1, 256, 192)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 64, 192)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 64, 208)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 64, 224)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 96, 160)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 96, 176)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 96, 192)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 96, 208)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 96, 224)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 128, 144)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 128, 160)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 128, 176)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 128, 192)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 256, 96)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 256, 112)
sub_iter.strides(0)[0] == 0INTERNAL ASSERT FAILED at "/pytorch/aten/src/ATen/native/cuda/Reduce.cuh":929, please report a bug to PyTorch. 
Configuration: (B, C, F, N) is (128, 1, 256, 128)
```

**Comment by leijue222** [comment]
This problem still occurs in torch 1.7.1. Only 8G is used for 24G memory.

**Comment by WMQ1999** [comment]
The problem happended in  1.8.0 to me....
When I use .cpu().sum().cuda() to replace .sum, it is solved.

**Comment by leijue222** [comment]
This seems to be a version problem of PyTorch, and I successfully circumvented the problem by using the 1.10.0 version of PyTorch.

**Comment by fingertap** [comment]
Same here on 1.8.0+cu111. Using GTX 3090, and it should not be a problem of memory size.

**Comment by leijue222** [comment]
> Same here on 1.8.0+cu111. Using GTX 3090, and it should not be a problem of memory size.

I solved the problem by upgrading pytorch to 1.10.0.

**Comment by fingertap** [comment]
> I solved the problem by upgrading pytorch to 1.10.0.

You are right. Updating to 1.10.0 solves the problem.
